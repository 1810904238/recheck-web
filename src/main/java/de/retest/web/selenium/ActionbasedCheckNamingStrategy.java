package de.retest.web.selenium;

import org.openqa.selenium.WebElement;

import com.google.common.collect.HashMultiset;
import com.google.common.collect.Multiset;

/**
 * This class aims to create semantic names for checks based on the previously executed action that led to that state.
 * This looks like "enter_[my text]_into_user" or "click_contact". If a name occurs multiple times, the second and
 * following occasions are labeled with "_2", "_3" etc. The drawback of this naming strategy is, that if e.g. the target
 * changes, the state will not be recognized anymore. So it is recommended to use this naming strategy _only_ in
 * combination with the <code>retestId</code>.
 */
public class ActionbasedCheckNamingStrategy implements AutocheckingCheckNamingStrategy {

	private final Multiset<String> checks = HashMultiset.create();

	@Override
	public String getUniqueCheckName( final String action, final WebElement targetElement, final Object... params ) {
		String result = action;
		// "enter_[" + normalizeAndShorten( keysToSend ) + "]_into"
		if ( "enter".equals( action ) ) {
			// TODO Call FileUtils.normalize
			result = "enter_[" + shortenTextInput( (CharSequence[]) params ) + "]_into";
		}
		if ( "get".equals( action ) ) {
			// TODO Call FileUtils.normalize
			result = "get_[" + shortenUrl( params[0] ) + "]";
		}
		if ( targetElement != null ) {
			result = result + "_" + toString( targetElement );
		}
		return makeUnique( result );
	}

	protected String makeUnique( final String result ) {
		checks.add( result );
		if ( checks.count( result ) == 1 ) {
			return result;
		}
		return result + "_" + checks.count( result );
	}

	protected String toString( final WebElement targetElement ) {
		String result = targetElement.toString();
		if ( result.contains( "->" ) ) {
			// remove driver info
			result = result.substring( result.lastIndexOf( "->" ) + 2 ).trim();
			// remove trailing ]
			result = result.substring( 0, result.length() - 1 );
			// remove identification criterion (e.g. id, class, ...)
			result = result.substring( result.lastIndexOf( ':' ) + 1 ).trim();
		}
		return result;
	}

	protected String shortenTextInput( final CharSequence[] keysToSend ) {
		if ( keysToSend == null || keysToSend.length == 0 || keysToSend[0] == null || keysToSend[0].length() == 0 ) {
			// How to properly represent empty string?
			return "";
		}
		final String stringToSend = keysToSend[0].toString();
		if ( stringToSend.length() <= 10 ) {
			return stringToSend;
		}
		return stringToSend.substring( 0, 7 ) + "...";
	}

	protected String shortenUrl( final Object url ) {
		if ( url == null ) {
			return "";
		}
		return url.toString().replace( "http://", "" ).replace( "https://", "" );
	}

	@Override
	public void nextTest() {
		checks.clear();
	}
}
