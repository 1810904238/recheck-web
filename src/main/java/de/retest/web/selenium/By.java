package de.retest.web.selenium;

import java.util.List;
import java.util.function.Predicate;

import de.retest.ui.descriptors.Element;
import de.retest.ui.descriptors.RootElement;
import de.retest.ui.diff.Alignment;

public abstract class By extends org.openqa.selenium.By {

	public static ByRetestId retestId( final String retestId ) {
		return new ByRetestId( retestId );
	}

	public static Element findElement( final RootElement lastExpectedState, final RootElement lastActualState,
			final Predicate<Element> predicate ) {
		if ( lastExpectedState == null ) {
			throw new IllegalArgumentException( "Cannot find element in null state." );
		}
		final Element result = findElement( lastExpectedState.getContainedElements(), predicate );
		if ( result == null ) {
			return null;
		}
		// TODO Use unobfuscated methods with retest 3.1.0
		// final Alignment alignment = Alignment.createAlignment( lastExpectedState, lastActualState );
		final Alignment alignment = Alignment.a( lastExpectedState, lastActualState );
		// return new Mapping<>( result, alignment.get( result ) );
		return alignment.a( result );
	}

	private static Element findElement( final List<Element> children, final Predicate<Element> predicate ) {
		for ( final Element element : children ) {
			if ( predicate.test( element ) ) {
				return element;
			}
			final Element result = findElement( element.getContainedElements(), predicate );
			if ( result != null ) {
				return result;
			}
		}
		return null;
	}

	public static Element findElementByAttribute( final RootElement lastExpectedState,
			final RootElement lastActualState, final String attributeName, final Predicate<Object> condition ) {
		return findElement( lastExpectedState, lastActualState, element -> {
			if ( element.getIdentifyingAttributes().get( attributeName ) != null ) {
				return condition.test( element.getIdentifyingAttributes().getAttribute( attributeName ).getValue() );
			} else if ( element.getAttributes().get( attributeName ) != null ) {
				return condition.test( element.getAttributes().get( attributeName ) );
			} else {
				return false;
			}
		} );
	}

	public static Element findElementByAttribute( final RootElement lastExpectedState,
			final RootElement lastActualState, final String attributeName, final Object attributeValue ) {
		return findElementByAttribute( lastExpectedState, lastActualState, attributeName, attributeValue::equals );
	}

}
