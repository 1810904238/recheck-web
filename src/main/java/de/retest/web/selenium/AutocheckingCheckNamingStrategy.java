package de.retest.web.selenium;

import org.openqa.selenium.WebElement;

/**
 * Interface to create names for checks for the {@link AutocheckingRecheckDriver}. These names must be unique within a
 * test, but can repeat in between tests (i.e. after calling {@link #nextTest()}).
 */
public interface AutocheckingCheckNamingStrategy {

	/**
	 * Return a string that is a unique identifier in the current test.
	 */
	String getUniqueCheckName( String action, WebElement target, Object... params );

	/**
	 * Reset this naming strategy, such that strings can be reused in different tests.
	 */
	void nextTest();
}
