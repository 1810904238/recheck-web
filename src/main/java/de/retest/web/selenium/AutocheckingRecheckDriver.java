package de.retest.web.selenium;

import java.util.List;
import java.util.stream.Collectors;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.remote.RemoteWebDriver;

import de.retest.recheck.RecheckOptions;
import de.retest.web.RecheckWebImpl;
import de.retest.web.RecheckWebOptions;

/**
 * Extends {@link UnbreakableDriver} and automagically creates a check <em>after</em> any executed action (usually
 * starting with {@link WebDriver#get(String)}). Consequently, you can omit using a {@code Recheck} instance such as
 * {@code RecheckImpl} or {@code RecheckWebImpl} instance. It utilizes the given {@link AutocheckingCheckNamingStrategy}
 * to create names for the checks.
 */
public class AutocheckingRecheckDriver extends UnbreakableDriver {

	private RecheckWebImpl re;
	private final RecheckOptions options;
	private final AutocheckingCheckNamingStrategy checkNamingStrategy;

	public AutocheckingRecheckDriver( final RemoteWebDriver wrapped ) {
		this( wrapped, RecheckWebOptions.builder().build() );
	}

	/**
	 * @deprecated use {@link #AutocheckingRecheckDriver(RemoteWebDriver, RecheckWebOptions)} instead.
	 *
	 * @param wrapped
	 *            The {@link RemoteWebDriver} to wrap.
	 * @param options
	 *            The {@link RecheckOptions} to use.
	 */
	@Deprecated
	public AutocheckingRecheckDriver( final RemoteWebDriver wrapped, final RecheckOptions options ) {
		super( wrapped );
		this.options = options;
		checkNamingStrategy = RecheckWebOptions.builder().build().getCheckNamingStrategy();
	}

	/**
	 * @deprecated use {@link #AutocheckingRecheckDriver(RemoteWebDriver, RecheckWebOptions)} instead.
	 *
	 * @param wrapped
	 *            The {@link RemoteWebDriver} to wrap.
	 * @param options
	 *            The {@link RecheckOptions} to use.
	 * @param namingStrategy
	 *            The {@link AutocheckingCheckNamingStrategy} to use.
	 */
	@Deprecated
	public AutocheckingRecheckDriver( final RemoteWebDriver wrapped, final RecheckOptions options,
			final AutocheckingCheckNamingStrategy namingStrategy ) {
		super( wrapped );
		this.options = options;
		checkNamingStrategy = RecheckWebOptions.builder().build().getCheckNamingStrategy();
	}

	public AutocheckingRecheckDriver( final RemoteWebDriver wrapped, final RecheckWebOptions options ) {
		super( wrapped );
		this.options = options;
		checkNamingStrategy = options.getCheckNamingStrategy();
	}

	public void startTest() {
		checkNamingStrategy.nextTest();
		if ( re == null ) {
			re = new RecheckWebImpl( options );
		}
		re.startTest();
	}

	public void startTest( final String testName ) {
		if ( re == null ) {
			re = new RecheckWebImpl( options );
		}
		re.startTest( testName );
	}

	public void capTest() {
		re.capTest();
	}

	public void cap() {
		checkNamingStrategy.nextTest();
		re.cap();
	}

	@Override
	public void get( final String url ) {
		super.get( url );
		check( "get", null );
	}

	@Override
	public void close() {
		// Is this sensible? What about tests using separate sessions?
		cap();
		super.close();
	}

	@Override
	public void quit() {
		try {
			// Is this sensible? What about tests using separate sessions?
			cap();
		} finally {
			super.quit();
		}
	}

	@Override
	public WebElement findElement( final ByBestMatchToRetestId by ) {
		final WebElement wrapped = super.findElement( by );
		String result = wrapped.toString();
		// replace " -> xpath: HTML[1]/BODY[1]/A[1]"
		result = result.substring( 0, result.lastIndexOf( " -> xpath: " ) );
		final String representation = result + " -> retestId: " + by.getRetestId() + "]";
		return new WebElementWrapper( wrapped, this ) {
			@Override
			public String toString() {
				return representation;
			}
		};
	}

	@Override
	public WebElement findElement( final By by ) {
		final WebElement result = super.findElement( by );
		if ( result instanceof WebElementWrapper ) {
			// Element was not found, so we already have it wrapped
			return result;
		}
		return new WebElementWrapper( result, this );
	}

	@Override
	public List<WebElement> findElements( final By by ) {
		return super.findElements( by ).stream() //
				.map( element -> new WebElementWrapper( element, this ) ) //
				.collect( Collectors.toList() );
	}

	void check( final String action, final WebElement target, final Object... params ) {
		if ( re == null ) {
			startTest();
		}
		re.check( this, checkNamingStrategy.getUniqueCheckName( action, target, params ) );
	}

	void check( final String action ) {
		if ( re == null ) {
			startTest();
		}
		re.check( this, checkNamingStrategy.getUniqueCheckName( action ) );
	}

	@Override
	public Navigation navigate() {
		return new AutocheckingNavigationWrapper( super.navigate(), this );
	}
}
