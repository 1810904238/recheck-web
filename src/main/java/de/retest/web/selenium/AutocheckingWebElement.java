package de.retest.web.selenium;

import java.util.List;
import java.util.stream.Collectors;

import org.openqa.selenium.By;
import org.openqa.selenium.Dimension;
import org.openqa.selenium.OutputType;
import org.openqa.selenium.Point;
import org.openqa.selenium.Rectangle;
import org.openqa.selenium.WebDriverException;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.WrapsElement;

public class AutocheckingWebElement implements WebElement, WrapsElement {

	private final WebElement wrappedElement;
	private final AutocheckingRecheckDriver driver;

	public AutocheckingWebElement( final WebElement wrappedElement, final AutocheckingRecheckDriver driver ) {
		if ( wrappedElement instanceof AutocheckingWebElement ) {
			throw new IllegalArgumentException( "Cannot wrap WebElementWrapper inside WebElementWrapper." );
		}
		this.wrappedElement = wrappedElement;
		this.driver = driver;
	}

	/**
	 * Skip checks for actions performed on this web element.
	 *
	 * @return returns the plain WebElement
	 */
	public WebElement skipCheck() {
		return wrappedElement;
	}

	@Override
	public <X> X getScreenshotAs( final OutputType<X> target ) throws WebDriverException {
		return wrappedElement.getScreenshotAs( target );
	}

	@Override
	public void click() {
		wrappedElement.click();
		driver.check( "click", wrappedElement );
	}

	@Override
	public void submit() {
		wrappedElement.submit();
		driver.check( "submit", wrappedElement );
	}

	@Override
	public void sendKeys( final CharSequence... keysToSend ) {
		wrappedElement.sendKeys( keysToSend );
		driver.check( "enter", wrappedElement, (Object[]) keysToSend );
	}

	@Override
	public void clear() {
		wrappedElement.clear();
		driver.check( "clear", wrappedElement );
	}

	@Override
	public String getTagName() {
		return wrappedElement.getTagName();
	}

	@Override
	public String getAttribute( final String name ) {
		return wrappedElement.getAttribute( name );
	}

	@Override
	public boolean isSelected() {
		return wrappedElement.isSelected();
	}

	@Override
	public boolean isEnabled() {
		return wrappedElement.isEnabled();
	}

	@Override
	public String getText() {
		return wrappedElement.getText();
	}

	@Override
	public List<WebElement> findElements( final By by ) {
		return wrappedElement.findElements( by ).stream() //
				.map( element -> new AutocheckingWebElement( element, driver ) ) //
				.collect( Collectors.toList() );
	}

	@Override
	public WebElement findElement( final By by ) {
		return new AutocheckingWebElement( wrappedElement.findElement( by ), driver );
	}

	@Override
	public boolean isDisplayed() {
		return wrappedElement.isDisplayed();
	}

	@Override
	public Point getLocation() {
		return wrappedElement.getLocation();
	}

	@Override
	public Dimension getSize() {
		return wrappedElement.getSize();
	}

	@Override
	public Rectangle getRect() {
		return wrappedElement.getRect();
	}

	@Override
	public String getCssValue( final String propertyName ) {
		return wrappedElement.getCssValue( propertyName );
	}

	@Override
	public WebElement getWrappedElement() {
		return wrappedElement;
	}
}
