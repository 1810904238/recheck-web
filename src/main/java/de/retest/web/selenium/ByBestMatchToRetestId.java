package de.retest.web.selenium;

import java.io.Serializable;
import java.util.List;

import org.apache.commons.lang3.StringUtils;
import org.openqa.selenium.By;
import org.openqa.selenium.SearchContext;
import org.openqa.selenium.WebElement;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import de.retest.recheck.ui.descriptors.Element;
import de.retest.recheck.ui.descriptors.RootElement;

/**
 * ByBestMatchToRetestId will search for elements by their old retestId (the retestId they had in the Golden Master
 * state).
 *
 * Note that elements are assigned a retestId when they are created (i.e. also when the _new_ state is created). This
 * retestId might differ from the retestId the element had in the persisted (Golden Master) state. Therefore, all its
 * child elements have their respective _new_ (potentially different) retestId. Thus, the result of a call to
 * `findByRetestId` might return an element with children that have a different retestId than they have in the Golden
 * Master.
 */
public class ByBestMatchToRetestId extends By implements Serializable {

	private static final long serialVersionUID = -3787115401615364934L;

	private static final Logger logger = LoggerFactory.getLogger( ByBestMatchToRetestId.class );

	private final String retestId;

	public ByBestMatchToRetestId( final String retestId ) {
		this.retestId = retestId;
		if ( StringUtils.isBlank( retestId ) ) {
			throw new IllegalArgumentException( "retestId must not be empty." );
		}
	}

	@Override
	public List<WebElement> findElements( final SearchContext context ) {
		logger.info( "findElements called with {} of {}.", context, context.getClass() );
		throw new UnsupportedOperationException( "This class can only be used in conjunction with a RecheckDriver." );
	}

	/**
	 * Might return an element whose children have a different retestId than in the Golden Master.
	 */
	public Element findElement( final RootElement lastExpectedState, final RootElement lastActualState ) {
		final Element result = de.retest.web.selenium.By.findElement( lastExpectedState, lastActualState,
				element -> retestId.equals( element.getRetestId() ) );
		if ( result == null ) {
			throw new NoElementWithReTestIdFoundException( retestId );
		}
		return result;
	}

	@Override
	public int hashCode() {
		return retestId.hashCode();
	}

	@Override
	public boolean equals( final Object other ) {
		if ( other == null ) {
			return false;
		}
		if ( this == other ) {
			return true;
		}
		if ( getClass() != other.getClass() ) {
			return false;
		}
		return retestId.equals( ((ByBestMatchToRetestId) other).retestId );
	}

	public String getRetestId() {
		return retestId;
	}
}
