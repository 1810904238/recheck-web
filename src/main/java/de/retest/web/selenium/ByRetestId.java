package de.retest.web.selenium;

import java.io.Serializable;
import java.util.List;

import org.apache.commons.lang3.StringUtils;
import org.openqa.selenium.By;
import org.openqa.selenium.SearchContext;
import org.openqa.selenium.WebElement;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import de.retest.ui.descriptors.Element;
import de.retest.ui.descriptors.RootElement;
import de.retest.util.Mapping;

public class ByRetestId extends By implements Serializable {

	private static final long serialVersionUID = -3787115401615364934L;

	private static final Logger logger = LoggerFactory.getLogger( ByRetestId.class );

	private final String retestId;

	public ByRetestId( final String retestId ) {
		this.retestId = retestId;
		if ( StringUtils.isBlank( retestId ) ) {
			throw new IllegalArgumentException( "retestId must not be empty." );
		}
	}

	@Override
	public List<WebElement> findElements( final SearchContext context ) {
		logger.info( "findElements called with {} of {}.", context, context.getClass() );
		throw new UnsupportedOperationException( "This class can only be used in conjunction with a RecheckDriver." );
	}

	public Element findElement( final RootElement lastExpectedState, final RootElement lastActualState ) {
		final Mapping<Element, Element> oldNewMapping = de.retest.web.selenium.By.findElement( lastExpectedState,
				lastActualState, element -> retestId.equals( element.getRetestId() ) );
		final Element result = oldNewMapping.getValue();
		if ( result == null ) {
			throw new NoElementWithReTestIdFoundException( retestId );
		}
		return result;
	}

	@Override
	public int hashCode() {
		return retestId.hashCode();
	}

	@Override
	public boolean equals( final Object other ) {
		if ( other == null ) {
			return false;
		}
		if ( this == other ) {
			return true;
		}
		if ( getClass() != other.getClass() ) {
			return false;
		}
		return retestId.equals( ((ByRetestId) other).retestId );
	}
}
