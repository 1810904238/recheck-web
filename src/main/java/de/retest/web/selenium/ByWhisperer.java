package de.retest.web.selenium;

import java.lang.reflect.Field;

import org.openqa.selenium.By.ByClassName;
import org.openqa.selenium.By.ById;
import org.openqa.selenium.By.ByLinkText;
import org.openqa.selenium.By.ByName;

public class ByWhisperer {

	private ByWhisperer() {}

	public static String retrieveId( final ById by ) {
		try {
			final Field field = ById.class.getDeclaredField( "id" );
			field.setAccessible( true );
			return (String) field.get( by );
		} catch ( final ReflectiveOperationException e ) {
			throw new RuntimeException( "ById does not have an 'id' field?", e );
		}
	}

	public static String retrieveCSSClassName( final ByClassName by ) {
		try {
			final Field field = ByClassName.class.getDeclaredField( "className" );
			field.setAccessible( true );
			return (String) field.get( by );
		} catch ( final ReflectiveOperationException e ) {
			throw new RuntimeException( "ByClassName does not have a 'className' field?", e );
		}
	}

	public static String retrieveName( final ByName by ) {
		try {
			final Field field = ByName.class.getDeclaredField( "name" );
			field.setAccessible( true );
			return (String) field.get( by );
		} catch ( final ReflectiveOperationException e ) {
			throw new RuntimeException( "ByName does not have a 'name' field?", e );
		}
	}

	public static String retrieveLinkText( final ByLinkText by ) {
		try {
			final Field field = ByLinkText.class.getDeclaredField( "linkText" );
			field.setAccessible( true );
			return (String) field.get( by );
		} catch ( final ReflectiveOperationException e ) {
			throw new RuntimeException( "ByLinkText does not have a 'linkText' field?", e );
		}
	}
}
