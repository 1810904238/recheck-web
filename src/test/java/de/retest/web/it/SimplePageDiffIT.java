package de.retest.web.it;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.fail;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.openqa.selenium.WebDriver;

import de.retest.recheck.Recheck;
import de.retest.recheck.RecheckImpl;
import de.retest.web.testutils.PageFactory;
import de.retest.web.testutils.WebDriverFactory;
import de.retest.web.testutils.WebDriverFactory.Driver;

public class SimplePageDiffIT {

	WebDriver driver;
	Recheck re;

	@Before
	public void setup() {
		re = new RecheckImpl();
		driver = WebDriverFactory.driver( Driver.CHROME );
	}

	@Test
	public void testSimpleChange() throws Exception {
		re.startTest();

		driver.get( PageFactory.toPageUrlString( "simple-page-diff.html" ) );

		re.check( driver, "open" );
		try {
			re.capTest();
			fail( "Assertion Error expected" );
		} catch ( final AssertionError e ) {
			assertThat( e ).hasMessageContaining( "Test 'testSimpleChange' has 4 difference(s) in 1 state(s):" ) //
					.hasMessageEndingWith( "\tdiv at 'html[1]/body[1]/div[3]':\n" + //
							"\t\tid: expected=\"twoblocks\", actual=\"changedblock\"\n" + //
							"\tp [Some text] at 'html[1]/body[1]/div[3]/p[1]':\n" + //
							"\t\ttext: expected=\"Some text\", actual=\"Some changed text\"\n" + //
							"\tp [Some more text] at 'html[1]/body[1]/div[3]/p[2]':\n" + //
							"\t\twas deleted\n" + //
							"\th2 [Subheading] at 'html[1]/body[1]/h2[1]':\n" + //
							"\t\twas inserted" );
		}
	}

	@After
	public void tearDown() {
		driver.quit();
		re.cap();
	}
}
