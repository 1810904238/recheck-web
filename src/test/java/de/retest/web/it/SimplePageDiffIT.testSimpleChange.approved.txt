A detailed report will be created at '/home/travis/build/retest/recheck-web/target/test-classes/retest/recheck/de.retest.web.it.SimplePageDiffIT.report'. You can review the details by using our CLI (https://github.com/retest/recheck.cli/) or GUI (https://retest.de/review/).

1 check(s) in 'de.retest.web.it.SimplePageDiffIT' found the following difference(s):
Test 'testSimpleChange' has 4 difference(s) in 1 state(s):
open resulted in:
	Metadata Differences:
		browser.name: expected="null", actual="chrome"
		browser.version: expected="null", actual="79.0.3945.130"
		check.type: expected="null", actual="driver"
		driver.type: expected="null", actual="ChromeDriver"
		os.name: expected="null", actual="LINUX"
		os.version: expected="null", actual="5.0.0-1026-gcp"
		url: expected="null", actual="file:///home/travis/build/retest/recheck-web/src/test/resources/pages/simple-page-diff.html"
		window.height: expected="null", actual="800"
		window.width: expected="null", actual="1200"
	div at 'html[1]/body[1]/div[3]':
		id: expected="twoblocks", actual="changedblock"
	p [Some text] at 'html[1]/body[1]/div[3]/p[1]':
		text: expected="Some text", actual="Some changed text"
	p [Some more text] at 'html[1]/body[1]/div[3]/p[2]':
		was deleted
	h2 [Subheading] at 'html[1]/body[1]/h2[1]':
		was inserted
